massa-network/src/peer_info_database.rs

```diff
@@ -847,47 +768,345 @@ impl PeerInfoDatabase {
                     Err(NetworkConnectionErrorType::MaxPeersConnectionReached(*ip))
                 } else {
                     peer.active_in_connections += 1;
-                    if peer.bootstrap {
-                        self.active_bootstrap_connections += 1;
-                    } else {
-                        self.active_in_nonbootstrap_connections += 1;
-                    }
                     Ok(())
                 }
             }
             hash_map::Entry::Vacant(vac) => {
-                let mut peer = PeerInfo {
-                    ip: *ip,
-                    banned: false,
-                    bootstrap: false,
-                    last_alive: None,
-                    last_failure: None,
-                    advertised: false,
-                    active_out_connection_attempts: 0,
-                    active_out_connections: 0,
-                    active_in_connections: 0,
-                };
-                if self.active_in_nonbootstrap_connections
-                    >= self.network_settings.max_in_nonbootstrap_connections
-                {
-                    Err(NetworkConnectionErrorType::MaxPeersConnectionReached(*ip))
-                } else if peer.active_in_connections
-                    >= self.network_settings.max_in_connections_per_ip
-                {
+                let mut peer = PeerInfo::new(*ip, false);
+                if peer.active_in_connections >= self.network_settings.max_in_connections_per_ip {
                     self.request_dump()?;
                     Err(NetworkConnectionErrorType::MaxPeersConnectionReached(*ip))
                 } else {
                     peer.active_in_connections += 1;
                     vac.insert(peer);
-                    self.active_in_nonbootstrap_connections += 1;
                     Ok(())
                 }
             }
         };
+        let peer = self.get_peer_type(ip).ok_or({
+            NetworkError::PeerConnectionError(NetworkConnectionErrorType::PeerInfoNotFoundError(
+                *ip,
+            ))
+        })?;
+
+        self.update_global_active_in_connection_count(
+            peer,
+            true,
+            NetworkConnectionErrorType::UnexpectedError,
+        )?;
         if let Err(res) = res {
             return Err(NetworkError::PeerConnectionError(res));
         }
         self.request_dump()?;
         Ok(())
     }
+
+    ////////////////////
+    // public getters //
+    ////////////////////
+
+    /// Sorts peers by ( last_failure, rev(last_success) )
+    /// and returns as many peers as there are available slots to attempt outgoing connections to.
+    pub fn get_out_connection_candidate_ips(&self) -> Result<Vec<IpAddr>, NetworkError> {
+        let mut res: Vec<_> = self
+            .get_candidate_ips_for_type(
```

Comment:
let's call it get_out_connection_candidate_ips_for_type for consistency (yes it's a bit long but makes it clearer :))